### PR TITLE
REALMC-9460: Prevent template app creation prompt from occurring automatically

### PR DESCRIPTION
**Pull request by andrewzlchen**
## Description

In this PR, I've reworked how the template app creation works within the realm-cli.

We decided offline that we will start out without making a prompt for template ids and instead add the available template ids   to the help menu. We will update the help menu as we add more templates and rethink how we want to prompt for template ids once we find a good solution to REALMC-9512.

~First off, I've updated the --template flag to have a default value when it is declared but is not supplied any arguments.~

~This allows us to distinguish between the following commands:~

~1. `realm-cli app create --template=<template-id>`~

~2. `realm-cli app create --template`~

~3. `realm-cli app create`~

~1 does not prompt the user for a template because it's already supplied~
~2 prompts the user for a template~
~3 does not prompt the user for template because we assume that users do not always want to create a template app.~

~## Updated Help menu~

image

~`<prompt-templates>` is a special value used to distinguish whether we should prompt the user. This is probably going to be the point of contention in this PR since we don't really have this behavior anywhere else in the CLI. I'm happy to try and work this out and iron out the details if this is not acceptable.~

### CODE REVIEW

File: internal/commands/app/create.go [hunk 1]
```diff
@@ -20,6 +20,10 @@ import (
 	"github.com/briandowns/spinner"
 )
 
+var (
+	noArgsDefaultValueTemplate = "<prompt-templates>"
```
Comment (andrewzlchen) [inline internal/commands/app/create.go:24]: this is the default value of the --template flag when a user declares it without any args

File: internal/commands/app/create.go [hunk 2]
```diff
@@ -94,11 +101,13 @@ func (cmd *CommandCreate) Flags() []flags.Flag {
...
 		},
 		flags.StringFlag{
-			Value: &cmd.inputs.Template,
+			Value:              &cmd.inputs.Template,
+			NoArgsDefaultValue: &noArgsDefaultValueTemplate,
 			Meta: flags.Meta{
 				Name: flagTemplate,
 				Usage: flags.Usage{
 					Description: "Create your Realm app from an available template",
+					Note:        "Manually specifying a template id must be done with a '='. ie: --template=web.graphql.todo",
```
Comment (andrewzlchen) [inline internal/commands/app/create.go:110]: this is the limitation of using `NoArgsDefaultValue`

https://github.com/spf13/pflag/issues/134

In practice, I think a relatively few amount of people would encounter this, but I'm hoping that this note is enough to cover us for the portion of users that know the template id and would like to specify them manually

File: internal/commands/app/create.go [hunk 3]
```diff
@@ -160,7 +169,7 @@ func (cmd *CommandCreate) Handler(profile *user.Profile, ui terminal.UI, clients
 		return err
 	}
 
-	if err = cmd.inputs.resolveTemplateID(ui, clients.Realm); err != nil {
+	if err := cmd.inputs.resolveTemplateID(ui, clients.Realm); err != nil {
```
Comment (andrewzlchen) [inline internal/commands/app/create.go:172]: driveby to control variable scoping to avoid weird bugs

File: internal/commands/app/create_inputs.go [hunk 4]
```diff
@@ -322,7 +329,11 @@ func (i createInputs) args(omitDryRun bool) []flags.Arg {
 		args = append(args, flags.Arg{flagLocalPathCreate, i.LocalPath})
 	}
 	if i.Template != "" {
-		args = append(args, flags.Arg{flagTemplate, i.Template})
+		if i.Template == noArgsDefaultValueTemplate {
+			args = append(args, flags.Arg{Name: flagTemplate})
```
Comment (andrewzlchen) [inline internal/commands/app/create_inputs.go:333]: remove `<prompt-template>` when we display the command if a user declared --template with no args

File: internal/commands/app/create.go (hunk 2 above)
Comment (UnicodeSnowman) [reply internal/commands/app/create.go:110]: ahh this is in fact a little awkward, but I do agree that users are not as likely to specify a template explicitly. what happens if they provide `--template web.graphql.todo`? do they get prompted to select a template? or does it complain about not knowing what `web.graphql.todo` means?

Comment (andrewzlchen) [reply internal/commands/app/create.go:110]: The current behavior is that it ignores `web.graphql.todo` and prompts for a template id.

Comment (makesitgo) [reply internal/commands/app/create.go:110]: im confused by this as well, why does it ignore `web.graphql.todo`?  by ignore im guessing that means `inputs.Template` is blank still?

Comment (andrewzlchen) [reply internal/commands/app/create.go:110]: no the value of `inputs.Template` will be the default no-args value which is `<prompt-templates>`

in its current state,

`realm-cli app create --template web.graphql.todo`

is the same as

`realm-cli app create web.graphql.todo --template`

Comment (andrewzlchen) [reply internal/commands/app/create.go:110]: it seems like we don't process any keywords after `create` for the `realm-cli app` command so `web.graphql.todo` doesn't really do anything if you do `realm-cli app create --template web.graphql.todo`

Comment (makesitgo) [reply internal/commands/app/create.go:110]: alright managed to get to the bottom of this offline, I missed what `NoArgsDefaultValue` was entailing at first.  I think we can ultimately avoid this awkwardness while still achieving the same functionality by tackling [REALMC-9512](https://jira.mongodb.org/browse/REALMC-9512) first.

spoke offline with andrew and he said he's happy to take on that ticket first

Comment (UnicodeSnowman) [reply internal/commands/app/create.go:110]: oh wow, interesting... that is indeed a bit hacky/funky, but it ends up working out okay.

**Comment by andrewzlchen** [comment]
blocked on REALMC-9512 for now. we're trying to figure out what our best course of action for providing a way to create a flag with optional arguments

File: internal/commands/app/create.go [hunk 5]
```diff
@@ -72,6 +73,7 @@ func (cmd *CommandCreate) Flags() []flags.Flag {
 				Name: flagClusterServiceName,
 				Usage: flags.Usage{
 					Description: "Specify the Realm app Service name to reference your Atlas cluster",
+					Note:        "Service names will be overwritten when creating a template app",
```
Comment (UnicodeSnowman) [inline internal/commands/app/create.go:76]: [q] I didn't catch/think about this until now, but should we stop referring to services here? should this just be "Data Source names..." and so on?

Comment (UnicodeSnowman) [reply internal/commands/app/create.go:76]: I feel like we can just change references to "service" here, I wouldn't consider that a breaking change... however, we probably _do_ need to leave the `cluster-service-name` flag/arg for now (to avoid a breaking change) and just add an additional `data-source` arg.

cc @makesitgo

**Review by UnicodeSnowman — APPROVED** [review]

File: internal/commands/app/create.go (hunk 5 above)
Comment (makesitgo) [reply internal/commands/app/create.go:76]: Agreed that changing the "note" text should be fine.  Do you mind elaborating on what the extra `data-source` arg would be?  Or are you suggesting replacing `cluster-service-name` with `cluster-data-source-name`?

Whatever we do with this one, there's an analogous `datalake-service-name` flag we should consider as well.  And if we're in the mood for breaking changes, maybe we remove the support for "multiple data sources at once", which would then eliminate the `*-service-name` flags all together (we could align with templates and go back to forcing `mongodb-atlas` as the name).

Comment (andrewzlchen) [reply internal/commands/app/create.go:76]: hmm I could be wrong, but maybe Chris is suggesting making 2 flags do the same thing: `--cluster-service-name` and `--cluster-data-source-name`. I'm hoping there's a way to alias flags so we don't need to manually account for that in our code if we end up going this route.

That said, I think this is a good opportunity to audit the CLI, Frontend (and maybe backend) to find wherever we still refer to data sources or HTTP endpoints as services and change them accordingly.

I can make a new ticket for this if we want to do this

Comment (andrewzlchen) [reply internal/commands/app/create.go:76]: https://jira.mongodb.org/browse/REALMC-9950

File: internal/commands/app/create_inputs.go [hunk 6]
```diff
@@ -174,7 +174,16 @@ func (i *createInputs) resolveClusters(ui terminal.UI, client atlas.Client, grou
...
 		var clusterName string
 		if len(i.Clusters) == 1 {
-			clusterName = i.Clusters[0]
+			for _, c := range clusters {
+				if c.Name == i.Clusters[0] {
+					clusterName = i.Clusters[0]
+				}
+			}
+			if clusterName == "" {
+				return nil, nil, fmt.Errorf("could not find atlas cluster '%s'", i.Clusters[0])
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:183]:
```suggestion
				return nil, nil, fmt.Errorf("could not find Atlas cluster '%s'", i.Clusters[0])
```

File: internal/commands/app/create_inputs.go [hunk 7]
```diff
@@ -174,7 +174,16 @@ func (i *createInputs) resolveClusters(ui terminal.UI, client atlas.Client, grou
...
+			}
+		} else if len(clusters) == 1 {
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:185]: do we have a test case for this already?  can we add one if not?

File: internal/commands/app/create_test.go [hunk 8]
```diff
@@ -335,75 +330,59 @@ Check out your app: cd ./test-app && realm-cli app describe
...
-			return &clientZipPkg.Reader, true, nil
-		}
-		ac := mock.AtlasClient{}
-		ac.GroupsFn = func() ([]atlas.Group, error) {
-			return []atlas.Group{{ID: "123"}}, nil
-		}
-		ac.ClustersFn = func(groupID string) ([]atlas.Cluster, error) {
-			return []atlas.Cluster{{Name: "test-cluster"}}, nil
+
+		client := mock.RealmClient{
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:343]: [nit] can we rename this to `realmClient` and define `atlasClient` similarly alongside here (rather than inline to the `cmd.Handler` call below)?

File: internal/commands/app/create_test.go [hunk 9]
```diff
@@ -335,75 +330,59 @@ Check out your app: cd ./test-app && realm-cli app describe
...
 				DeploymentModel: realm.DeploymentModelGlobal,
-				ConfigVersion:   realm.DefaultAppConfigVersion,
 			},
 			Clusters: []string{"test-cluster"},
 		}}
 
-		assert.Nil(t, cmd.Handler(profile, ui, cli.Clients{Realm: rc, Atlas: ac}))
-
-		console.Tty().Close() // flush the writers
-		<-doneCh              // wait for procedure to complete
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:400]: is this test no longer interactive (i.e. is `procedure` needed at all)?

if its no longer interactive, we should switch the `ui` to just be `mock.NewUI` (rather than the `V10XConsole` variant), as well as this bit:

```go
		doneCh := make(chan (struct{}))
		go func() {
			defer close(doneCh)
			procedure(console)
		}()
```

if it _is_ interactive still, then we definitely should still keep the `<-doneCh` and psuedo-terminal `Close` calls

File: internal/commands/app/create_test.go [hunk 10]
```diff
@@ -335,75 +330,59 @@ Check out your app: cd ./test-app && realm-cli app describe
...
 
-		path := filepath.Join(profile.WorkingDirectory, cmd.inputs.Name, local.BackendPath)
-		appLocal, err := local.LoadApp(path)
+		_, err = local.LoadApp(filepath.Join(profile.WorkingDirectory, cmd.inputs.Name))
 		assert.Nil(t, err)
 
-		assert.Equal(t, appLocal.RootDir, path)
+		backendFileInfo, err := ioutil.ReadDir(filepath.Join(profile.WorkingDirectory, cmd.inputs.Name))
+		assert.Nil(t, err)
+		assert.Equal(t, len(backendFileInfo), 10)
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:385]: was this an existing assert?  do you mind at least providing some clarity into where the `10` comes from?  (and/or is there a better/more specific assert we can use instead of this?)

File: internal/commands/app/new_app_inputs.go [hunk 11]
```diff
@@ -38,22 +36,21 @@ func (i *newAppInputs) resolveRemoteApp(ui terminal.UI, rc realm.Client) (realm.
...
 
-func (i *newAppInputs) resolveTemplateID(ui terminal.UI, client realm.Client) error {
-	if i.Template == "" && ui.AutoConfirm() {
+// resolveTemplateID is responsible for resolving a template id from a user cli request.
+// a user can use the --template flag in 1 of the following ways:
+// 1. --template=<template-id>: This populates the value of i.Template to be <template-id> and attempts to create a
+//    template with id = <template-id>
+// 2. no flag defined: This populates the value of i.Template to be "". This tells us to skip the template-app-related
+//    parts of app creation and create a default app
+func (i *newAppInputs) resolveTemplateID(client realm.Client) error {
```
Comment (makesitgo) [inline internal/commands/app/new_app_inputs.go:45]: I might simplify this docstring a bit (I don't think we need much explanation here about the way flags work/set values)

```go
// resolveTemplateID is responsible for resolving a template id from a user cli request.
// If the --template flag is not set, the CLI will create a default app without any template,
// otherwise the CLI will attempt to create an app based on the specified template
func (i *newAppInputs) resolveTemplateID(client realm.Client) error {
```

File: internal/commands/pull/command.go [hunk 12]
```diff
@@ -84,13 +84,14 @@ func (cmd *Command) Flags() []flags.Flag {
...
-			Value: &cmd.inputs.TemplateID,
+		flags.StringSliceFlag{
+			Value: &cmd.inputs.TemplateIDs,
 			Meta: flags.Meta{
 				Name:      "template",
 				Shorthand: "t",
 				Usage: flags.Usage{
-					Description: "Specify the Template ID that is used for this Realm app",
+					Description: "Specify the IDs of the template frontend to export. Specified templates must be compatible with the remote app",
+					Note:        "Options include: web.graphql.todo, web.mql.todo, triggers, ios.swift.todo, android.kotlin.todo, react-native.todo, xamarin.todo",
```
Comment (makesitgo) [inline internal/commands/pull/command.go:94]: similar not here about using `AllowedValues` instead of `Note` for this data

File: internal/commands/pull/command.go [hunk 13]
```diff
@@ -84,13 +84,14 @@ func (cmd *Command) Flags() []flags.Flag {
...
-		flags.StringFlag{
-			Value: &cmd.inputs.TemplateID,
+		flags.StringSliceFlag{
+			Value: &cmd.inputs.TemplateIDs,
 			Meta: flags.Meta{
 				Name:      "template",
 				Shorthand: "t",
 				Usage: flags.Usage{
-					Description: "Specify the Template ID that is used for this Realm app",
+					Description: "Specify the IDs of the template frontend to export. Specified templates must be compatible with the remote app",
```
Comment (makesitgo) [inline internal/commands/pull/command.go:93]:
```suggestion
					Description: "Specify the frontend Template ID(s) to export",
					Note: "Specified templates must be compatible with the remote app",
```

File: internal/commands/pull/command_test.go [hunk 14]
```diff
@@ -478,18 +478,6 @@ Successfully pulled app down: app
 			assert.Nil(t, err)
 			defer console.Close()
 
-			doneCh := make(chan struct{})
```
Comment (makesitgo) [inline internal/commands/pull/command_test.go:481]: if this test is no longer interactive, then let's move the `ui` back to `mock.NewUI` (rather than the `V10X` variant)

File: internal/commands/pull/inputs.go [hunk 15]
```diff
@@ -27,7 +26,7 @@ type inputs struct {
 	IncludeDependencies bool
 	IncludeHosting      bool
 	DryRun              bool
-	TemplateID          string
+	TemplateIDs         []string
```
Comment (makesitgo) [inline internal/commands/pull/inputs.go:29]: [question] did Templates want to be an array input for `app create` similar to how it is here?

File: internal/commands/pull/inputs_test.go [hunk 16]
```diff
@@ -179,16 +179,24 @@ func TestPullTemplatesResolve(t *testing.T) {
 	assert.Nil(t, err)
 	defer templateZipPkg2.Close()
 
+	t.Run("should not do anything if no template id is passed in", func(t *testing.T) {
+		var realmClient mock.RealmClient
+
+		input := inputs{}
+		out, err := input.resolveClientTemplates(realmClient, "some-group-id", "some-app-id")
+		assert.Equal(t, nil, err)
```
Comment (makesitgo) [inline internal/commands/pull/inputs_test.go:187]: use `assert.Nil` instead

**Review by makesitgo — COMMENTED** [review]
nothing blocking from me, but just some nits/minor things

File: internal/commands/app/create.go (hunk 5 above)
Comment (andrewzlchen) [reply internal/commands/app/create.go:76]: Hmm, I'm not sure if this is a valid concern, but I think that depending on the solution we end up with in REALMC-9950, we may want to leave this flag alone.

I'd rather just keep the descriptions aligned with their flag name until we know for sure that the change is needed.

File: internal/commands/app/create.go [hunk 17]
```diff
@@ -99,6 +102,7 @@ func (cmd *CommandCreate) Flags() []flags.Flag {
 				Name: flagTemplate,
 				Usage: flags.Usage{
 					Description: "Create your Realm app from an available template",
+					Note:        "Valid templates: web.graphql.todo, web.mql.todo, triggers, ios.swift.todo, android.kotlin.todo, react-native.todo, xamarin.todo",
```
[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/commands/app/create.go:105]: ooo that's a good idea. I'll do that instead!

File: internal/commands/app/create.go (hunk 5 above)
Comment (makesitgo) [reply internal/commands/app/create.go:76]: no strong opinions on my end, I'd maybe just ask sumedha/drew if they want this quick update or not

File: internal/commands/app/create.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (makesitgo) [reply internal/commands/app/create.go:105]: btw I'm all ears if you have any thoughts on how to make these other fields more discoverable... slash how to make the `Meta`/`Usage` more fool-proof to use

File: internal/commands/app/create_inputs.go (hunk 7 above)
Comment (andrewzlchen) [reply internal/commands/app/create_inputs.go:185]: I've grouped these template app creation tests and added two new test cases:

1. For this comment: for whether a cluster will be automatically selected if user does not supply cluster and there is only one option returned from the atlas clusters endpoint
2. For whether a prompt appears if user does not supply cluster and there are multiple options returned from the atlas clusters endpoint

File: internal/commands/app/create_test.go (hunk 10 above)
Comment (andrewzlchen) [reply internal/commands/app/create_test.go:385]: ah yeah, that was an existing change that I updated. This 10 comes from the mock backend zip package which is the contents of `testdata/project.zip`.

This test serves to ensure that when a user does not supply --template, a default app will be created.

the assertion that the resulting directory has len = 10 is just to ensure that the result is the same as the result of the mocked export function.

maybe we can unzip this `testdata/project.zip` first and use this as the assertion value

ie:
```go
backendZipPkg, err := zip.OpenReader("testdata/project.zip")
assert.Nil(t, err)

// add this
expectedBackendZipLen := len(backendZipPkg.File)

realmClient := mock.RealmClient{
    ExportFn: func(groupID, appID string, req realm.ExportRequest) (string, *zip.Reader, error) {
        return "", &backendZipPkg.Reader, nil
    },
    // ...
}

// ...

// and use the new expected length here
assert.Equal(t, len(backendFileInfo), expectedBackendZipLen)
```

File: internal/commands/app/create_test.go (hunk 9 above)
Comment (andrewzlchen) [reply internal/commands/app/create_test.go:400]: i suppose it's because the structure of the test cases are similar, but the diff is trying to compare a deleted test case with a different one altogether

File: internal/commands/app/create_test.go (hunk 10 above)
Comment (andrewzlchen) [reply internal/commands/app/create_test.go:385]: edit nevermind, we cant do this. the zip file also lists out nested files so the number of files in the zip file != the number of `backendFileInfo` files.

I think it's probably sufficient to assert that:

1. we do not create a `backend` and `frontend` directory in the top level
2. and that realm_config.json exists in the top level to denote that this directory is a realm app folder

I'll do this for now, but I'm happy to change this if you have any opinions here

File: internal/commands/pull/inputs.go (hunk 15 above)
Comment (andrewzlchen) [reply internal/commands/pull/inputs.go:29]: User Story:

When you create a template, you choose a single template, but when you pull, you should have an option of what frontend templates you can create. I realize that this may be a bit confusing. Would calling this flag --template-client be better?

Explanation:

I think it was an oversight that this was just a string before, so I fixed it here. 

I didn't realize that this was a string until I tried to pull multiple templates without the prompt.

The prompt from before allowed users to pull multiple templates which I think we should maintain feature parity for.

File: internal/commands/app/create_inputs_test.go [hunk 18]
```diff
@@ -710,90 +710,166 @@ func TestAppCreateInputsResolveCluster(t *testing.T) {
 		assert.Equal(t, "123", expectedGroupID)
 	})
 
-	t.Run("should error when more than one cluster passed in when creating template app", func(t *testing.T) {
-		_, ui := mock.NewUI()
-		clusterNames := []string{"Cluster0", "Cluster1"}
+	t.Run("when creating template apps", func(t *testing.T) {
```
Comment (andrewzlchen) [inline internal/commands/app/create_inputs_test.go:713]: i decided to group all of the template app creation tests together to tidy them up a bit

Comment (andrewzlchen) [reply internal/commands/app/create_inputs_test.go:713]: hiding whitespace changes will make reading through this significantly easier

File: internal/commands/app/create_test.go [hunk 19]
```diff
@@ -335,75 +330,78 @@ Check out your app: cd ./test-app && realm-cli app describe
 			procedure(console)
 		}()
 
-		var createdApp realm.App
-		rc := mock.RealmClient{}
-
-		zipPkg, err := zip.OpenReader("testdata/project.zip")
-		assert.Nil(t, err)
-		defer zipPkg.Close()
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:343]: looks like we dropped the `defer Close()` on `defaultBackendZipPkg` (the same I think goes for `frontendZipPkg` in this file too?)

File: internal/commands/app/create_test.go [hunk 20]
```diff
@@ -335,75 +330,78 @@ Check out your app: cd ./test-app && realm-cli app describe
...
-		assert.Equal(t, appLocal.RootDir, path)
+		// we expect for the command to have created a default app. we will assert that realm_config.json exists and that
+		// backend/ and frontend/ do not exist in the directory.
+		var containsRealmConfig bool
+		var containsTemplateDirChild bool
+		for _, fi := range backendFileInfo {
+			if strings.Contains(fi.Name(), local.NameRealmConfig) {
+				containsRealmConfig = true
+			}
+			if fi.Name() == local.BackendPath || fi.Name() == local.FrontendPath {
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:395]: is it worth just using the same style of `strings.Contains` checks for these?

File: internal/commands/pull/command.go [hunk 21]
```diff
@@ -84,13 +84,14 @@ func (cmd *Command) Flags() []flags.Flag {
...
-		flags.StringFlag{
-			Value: &cmd.inputs.TemplateID,
+		flags.StringSliceFlag{
+			Value: &cmd.inputs.TemplateIDs,
 			Meta: flags.Meta{
 				Name:      "template",
 				Shorthand: "t",
 				Usage: flags.Usage{
-					Description: "Specify the Template ID that is used for this Realm app",
+					Description:   "Specify the frontend Template ID(s) to export. Specified templates must be compatible with the remote app",
```
Comment (makesitgo) [inline internal/commands/pull/command.go:93]: what are your thoughts on making the second sentence here a `Note`?  there's not a _ton_ of description/flag usages with periods/multiple sentences yet

**Review by makesitgo — APPROVED** [review]
last few bits here, but looks good.  nice work!

🆒  beans

File: internal/cloud/realm/templates.go [hunk 22]
```diff
@@ -11,19 +11,43 @@ import (
 	"github.com/10gen/realm-cli/internal/utils/api"
 )
 
+// AllowedTemplates is an array of templates that we support. We should update this slice whenever we add new templates.
```
[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/cloud/realm/templates.go:14]: just to confirm, is the 2nd line okay?

`// This array is used to display the allowed values in the --help menu of the templates flag`

File: internal/commands/app/create_test.go (hunk 20 above)
Comment (andrewzlchen) [reply internal/commands/app/create_test.go:395]: Hmm, I think I'd rather make the realm_config.json be an exact match than keeping the `strings.Contains` just so the assertion is as strict as possible.

I'll do that now but if you feel differently, I can change it

File: internal/cloud/realm/templates.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (makesitgo) [reply internal/cloud/realm/templates.go:14]: i'd remove that too, again not super important I just don't think it's all that necessary

File: internal/commands/pull/command.go (hunk 21 above)
Comment (andrewzlchen) [reply internal/commands/pull/command.go:93]: I'm okay with that but the only this is that the `Note` is not on it's own line.

It seems to be displayed after the `AllowedValues` so it may potentially be lost since the allowed values list is kinda long and not super human readable


image

File: internal/commands/app/create_test.go (hunk 20 above)
Comment (makesitgo) [reply internal/commands/app/create_test.go:395]: the exact match on `realm_config.json` is fine by me since ultimately the assert here is that `false` went to `true` (b/c this condition hit)

what worries me is the opposite style of assert for making sure we don't have `backend` or `frontend` directories.  if for w/e reason the `fi.Name()` started having a trailing slash or any other characters in it, this test wouldn't catch that (b/c `false` would have remained `false`).  that's why I'm wondering if it makes sense to at least use the more permissive `strings.Contains` check for this frontend/backend logic.

[question] instead of using `ioutil.ReadDir` on the whole project directory and iterating over its files, why don't we just explicitly look for `realm_config.json` (& verify it exists) and `backend`/`frontend` directories (& verify they don't exist)?

File: internal/commands/pull/command.go (hunk 21 above)
Comment (kpham1231) [reply internal/commands/pull/command.go:93]: Having the Note before the Defaults makes sense to me! Given the defaults can get pretty long, getting the Note across first would be good

File: internal/commands/app/create_test.go (hunk 20 above)
Comment (andrewzlchen) [reply internal/commands/app/create_test.go:396]: oh i see what you mean. That's true.

I think your suggestion here is good. I'll go ahead and do that
